### PR TITLE
Lazy write out

### PR DESCRIPTION
**Pull request by toolslive**
feature #302

### CODE REVIEW

File: ocaml/src/alba_bench.ml [hunk 1]
```diff
@@ -75,6 +75,7 @@ let do_scenarios
   Alba_client2.with_client
     ~tcp_keepalive:Tcp_keepalive2.default
     ~populate_osds_info_cache:true
+    ~upload_slack:0.2
```
Comment (toolslive) [inline ocaml/src/alba_bench.ml]: yes. will make it configurable for alba bench too

File: ocaml/src/maintenance_test.ml [hunk 2]
```diff
@@ -660,23 +760,27 @@ let test_verify_namespace () =
...
-             |> List.hd_exn
-             |> snd
-             = 1);
+     let new_version_missing = mf2.fragment_locations
+                               |> List.hd_exn
+                               |> List.hd_exn
+                               |> snd
+     in
+     OUnit.assert_bool "Manifest.version_id of formerly missing"
+       (1 <= new_version_missing);
```
Comment (domsj) [inline ocaml/src/maintenance_test.ml]: should be exactly equal to 1? same remark for the other comparison below

File: ocaml/src/alba_test.ml [hunk 3]
```diff
@@ -37,7 +37,16 @@ let test_with_alba_client ?bad_fragment_callback f =
         ~release_resources:true
         ~tcp_keepalive:Tcp_keepalive2.default
         ~populate_osds_info_cache:true
-        f
+        ~upload_slack:0.001
```
Comment (toolslive) [inline ocaml/src/alba_test.ml]: yes. yes.

File: ocaml/src/tools/lwt_extra2.ml [hunk 4]
```diff
@@ -298,3 +298,86 @@ let copy_between_fds fd_in fd_out size buffer =
   copy_using reader writer size buffer
 
 
+type 'a state =
```
Comment (toolslive) [inline ocaml/src/tools/lwt_extra2.ml]: I'll have a look to simplify this.

File: ocaml/src/alba_client_upload.ml [hunk 5]
```diff
@@ -419,18 +442,63 @@ let upload_object''
...
-           ~k ~m ~w'
+           ~k ~m ~w' ~min_fragment_count
            ~compression ~encryption ~fragment_checksum_algo
            ~version_id ~gc_epoch
            ~object_info_o
            ~osds:target_osds
            ~fragment_cache
-           ~cache_on_write)
+           ~cache_on_write
+           ~upload_slack
```
Comment (domsj) [inline ocaml/src/alba_client_upload.ml]: only add slack for the last chunk?
`~upload_slack:(if has_more then 0.0 else upload_slack`

Comment (toolslive) [inline ocaml/src/alba_client_upload.ml]: will do

File: ocaml/src/alba_osd.ml [hunk 6]
```diff
@@ -136,7 +136,8 @@ class client
                    ~object_id_hint:None
                    ~fragment_cache:(alba_client # get_base_client # get_fragment_cache)
                    ~cache_on_write:(alba_client # get_base_client # get_cache_on_read_write |> snd)
-                 >>= fun (mf, _chunk_fidmos, _, gc_epoch) ->
+                   ~upload_slack:0.2
```
Comment (domsj) [inline ocaml/src/alba_osd.ml]: upload_slack in this class should depend on what's configured in proxy config file too?

File: ocaml/src/nsm_client.ml [hunk 7]
```diff
@@ -150,9 +150,12 @@ class client (nsm_host_client : Nsm_host_client.basic_client) namespace_id =
...
         (allow_overwrite, object_name)
 
-    method update_manifest ~object_name ~object_id updated_object_locations ~gc_epoch ~version_id =
+    method update_manifest ~object_name ~object_id
+                           updated_object_locations
+                           ~gc_epoch ~version_id
+      =
       self # update
-        UpdateObject
+        UpdateObject2
```
Comment (toolslive) [inline ocaml/src/nsm_client.ml]: I'll adapt, but don't agree. Suppressing a call that no component is using is only hypothetically breaking backward compatibility.

File: ocaml/src/nsm_model.ml [hunk 8]
```diff
@@ -1836,6 +1904,31 @@ module NamespaceManager(C : Constants)(KV : Read_key_value_store) = struct
...
+        kv object_name object_id
+        (new_fragments : (chunk_id
+                          * fragment_id
+                          * osd_id option
+                          * (int * Checksum.t) option)
+                           list)
+        gc_epoch
+        version_id
+    =
+    failwith "not implemented yet"
```
Comment (toolslive) [inline ocaml/src/nsm_model.ml]: it did fail.

File: ocaml/src/alba_osd.ml (hunk 6 above)
Comment (toolslive) [inline ocaml/src/alba_osd.ml]: should it be the same as what's in the proxy's config ?

**Comment by domsj** [comment]
There are some changes needed for recover_nsm_host.ml too: it previously assumed fragment_sizes & fragment_checksums in the recovery info would be filled up with the actual correct values.
As of this PR it can no longer assume this to be the case.
I think there are 2 options for fixing this:
- get the fragment size/checksum from the asd (query metadata about some key - we currently don't have this sort of operation but we can add it, for backwards compatibility we can get the same by downloading the entire fragment)
- we could enforce that the fragment checksum/size in each recovery info object is correct for the fragment the recovery info is associated with, the recovery code can then use this...

(my preference goes towards the second option)

**Pull request merged**